File: ios/Sources/MqttQuicPlugin/Transport/StreamTransport.swift
//
// StreamTransport.swift
// MqttQuicPlugin
//
// Transport abstraction: StreamReader-like and StreamWriter-like.
// Phase 2 will implement these over QUIC streams; Phase 1 uses mocks for tests.
//

import Foundation

/// StreamReader-like interface: read(n), readexactly(n).
/// Async where applicable (Phase 2 uses async).
public protocol MQTTStreamReaderProtocol: AnyObject {
    func read(maxBytes: Int) async throws -> Data
    func readexactly(_ n: Int) async throws -> Data
}

/// StreamWriter-like interface: write(data), drain(), close().
public protocol MQTTStreamWriterProtocol: AnyObject {
    func write(_ data: Data) async throws
    func drain() async throws
    func close() async throws
}

// MARK: - Mock implementations (Phase 1 unit tests)

/// In-memory buffer for mock read/write.
public final class MockStreamBuffer {
    public private(set) var readBuffer: Data
    public private(set) var writeBuffer: Data
    public var isClosed: Bool = false

    public init(initialReadData: Data = Data()) {
        self.readBuffer = initialReadData
        self.writeBuffer = Data()
    }

    public func appendRead(_ data: Data) {
        readBuffer.append(data)
    }

    public func consumeRead(maxBytes: Int) -> Data {
        let n = min(maxBytes, readBuffer.count)
        let out = readBuffer.prefix(n)
        readBuffer = readBuffer.dropFirst(n)
        return Data(out)
    }

    public func appendWrite(_ data: Data) {
        writeBuffer.append(data)
    }

    public func consumeWrite() -> Data {
        let d = writeBuffer
        writeBuffer = Data()
        return d
    }
}

/// Mock reader: reads from MockStreamBuffer.
public final class MockStreamReader: MQTTStreamReaderProtocol {
    private let buffer: MockStreamBuffer

    public init(buffer: MockStreamBuffer) {
        self.buffer = buffer
    }

    public func read(maxBytes: Int) async throws -> Data {
        if buffer.isClosed && buffer.readBuffer.isEmpty {
            return Data()
        }
        let n = min(maxBytes, buffer.readBuffer.count)
        guard n > 0 else {
            try await Task.sleep(nanoseconds: 1_000_000) // 1ms
            return try await read(maxBytes: maxBytes)
        }
        return buffer.consumeRead(maxBytes: n)
    }

    public func readexactly(_ n: Int) async throws -> Data {
        var acc = Data()
        while acc.count < n {
            if buffer.isClosed { throw MQTTProtocolError.insufficientData("stream closed") }
            let chunk = try await read(maxBytes: n - acc.count)
            if chunk.isEmpty { throw MQTTProtocolError.insufficientData("readexactly") }
            acc.append(chunk)
        }
        return acc
    }
}

/// Mock writer: appends to MockStreamBuffer.
public final class MockStreamWriter: MQTTStreamWriterProtocol {
    private let buffer: MockStreamBuffer

    public init(buffer: MockStreamBuffer) {
        self.buffer = buffer
    }

    public func write(_ data: Data) async throws {
        buffer.appendWrite(data)
    }

    public func drain() async throws {}

    public func close() async throws {
        buffer.isClosed = true
    }
}

File: android/src/main/kotlin/ai/annadata/mqttquic/transport/StreamTransport.kt
package ai.annadata.mqttquic.transport

import kotlinx.coroutines.delay

/**
 * StreamReader-like: read(n), readexactly(n).
 * Phase 2 implements over QUIC stream.
 */
interface MQTTStreamReader {
    /** Number of bytes currently buffered (without reading from stream). Used to skip unwanted packets only when safe. */
    suspend fun available(): Int
    suspend fun read(maxBytes: Int): ByteArray
    suspend fun readexactly(n: Int): ByteArray
}

/**
 * StreamWriter-like: write(data), drain(), close().
 */
interface MQTTStreamWriter {
    suspend fun write(data: ByteArray)
    suspend fun drain()
    suspend fun close()
}

/**
 * In-memory buffer for mock read/write (Phase 1 unit tests).
 */
class MockStreamBuffer(initialReadData: ByteArray = ByteArray(0)) {
    var readBuffer = initialReadData.toMutableList()
        private set
    val writeBuffer = mutableListOf<Byte>()
    var isClosed = false
        private set

    fun appendRead(data: ByteArray) {
        readBuffer.addAll(data.toList())
    }

    fun consumeWrite(): ByteArray {
        val d = writeBuffer.toByteArray()
        writeBuffer.clear()
        return d
    }

    fun close() {
        isClosed = true
    }
}

/**
 * Mock reader over MockStreamBuffer.
 */
class MockStreamReader(private val buffer: MockStreamBuffer) : MQTTStreamReader {

    override suspend fun available(): Int = buffer.readBuffer.size

    override suspend fun read(maxBytes: Int): ByteArray {
        if (buffer.isClosed && buffer.readBuffer.isEmpty()) return ByteArray(0)
        val n = minOf(maxBytes, buffer.readBuffer.size)
        if (n == 0) {
            delay(1)
            return read(maxBytes)
        }
        val out = buffer.readBuffer.take(n).toByteArray()
        repeat(n) { buffer.readBuffer.removeAt(0) }
        return out
    }

    override suspend fun readexactly(n: Int): ByteArray {
        val acc = mutableListOf<Byte>()
        while (acc.size < n) {
            if (buffer.isClosed) throw IllegalArgumentException("stream closed")
            val chunk = read(n - acc.size)
            if (chunk.isEmpty()) throw IllegalArgumentException("readexactly")
            acc.addAll(chunk.toList())
        }
        return acc.toByteArray()
    }
}

/**
 * Mock writer over MockStreamBuffer.
 */
class MockStreamWriter(private val buffer: MockStreamBuffer) : MQTTStreamWriter {

    override suspend fun write(data: ByteArray) {
        buffer.writeBuffer.addAll(data.toList())
    }

    override suspend fun drain() {}

    override suspend fun close() {
        buffer.close()
    }
}
